api/learn/learn.go

```diff
@@ -1,22 +1,143 @@
+)
 
 // API is the exported APIClient, it is set during Init
 var API *APIClient
 
 // APIClient makes network API calls to Learn
 type APIClient struct {
-	client  api.Client
-	token   string
-	baseURL string
+	client      api.Client
+	baseURL     string
+	Credentials *Credentials
+}
+
+// Credentials represents the shape of data that the initial call to Learn
+// for s3 and slack credentials will hydrate
+type Credentials struct {
+	*APIToken         `json:"api_token"`
+	*S3Credentials    `json:"s3_credentials"`
+	*SlackCredentials `json:"slack_credentials"`
+}
+
+// S3Credentials represents the important AWS credentials we retrieve from Learn
+// with an api_token
+type S3Credentials struct {
+	AccessKeyID     string `json:"access_key_id"`
+	SecretAccessKey string `json:"secret_access_key"`
+	KeyPrefix       string `json:"key_prefix"`
+	BucketName      string `json:"bucket_name"`
+}
+
+// SlackCredentials represents the credentials we retrieve from Learn for the CLI
+// to operate correctly
+type SlackCredentials struct {
+	DevNotifyURL string `json:"dev_notify_url"`
+}
+
+// APIToken is a simple wrapper around an API token
+type APIToken struct {
+	token string
+}
+
+// CredentialsResponse describes the shape of the return data from the call
+// to RetrieveCredentials
+type CredentialsResponse struct {
+	S3    S3Credentials    `json:"s3"`
+	Slack SlackCredentials `json:"slack"`
 }
 
 // NewAPI is a constructor for the ApiClient
-func NewAPI(token, baseURL string, client api.Client) *APIClient {
-	return &APIClient{
+func NewAPI(baseURL string, client api.Client) (*APIClient, error) {
+	apiClient := &APIClient{
 		client:  client,
-		token:   token,
 		baseURL: baseURL,
 	}
+
+	// Retrieve the application credentials for the CLI using a user's API token
+	creds, err := apiClient.RetrieveCredentials()
+	if err != nil {
+		return nil, errors.New("Could not retrieve credentials from Learn. Please ensure you have the right API token in your ~/.glearn-config.yaml")
+	}
+
+	apiClient.Credentials = creds
+
+	return apiClient, nil
+}
+
+// RetrieveCredentials uses a user's api_token to request AWS credentials
+// from Learn. It returns a populated *S3Credentials struct or an error
+func (api *APIClient) RetrieveCredentials() (*Credentials, error) {
+	// Early return if user's api_token is not set
+	apiToken, ok := viper.Get("api_token").(string)
+	if !ok {
+		return nil, errors.New("Please set your api_token in ~/.glearn-config.yaml")
+	}
+
+	req, err := http.NewRequest("GET", fmt.Sprintf("%s/api/v1/users/learn_cli_credentials", api.baseURL), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", apiToken))
+
+	res, err := api.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	var c CredentialsResponse
+
+	err = json.NewDecoder(res.Body).Decode(&c)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Credentials{
```

Comment:
Could this be accomplished by setting the api token directly on the `c CredentialsResponse` object? It will have already unmarshaled the json for S3Credentials and SlackCredentials values

Reply:
Nvm I see they're pointer values on the struct. Out of curiosity is that just to pass by reference the credentials?